<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<title>- OptiMAS - decision support for marker-assisted selection</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rel="stylesheet" media="screen,projection" type="text/css" href="css/reset.css" />
<link rel="stylesheet" media="screen,projection" type="text/css" href="css/main.css" />
<link rel="stylesheet" media="screen,projection" type="text/css" href="css/2col.css" title="2col" />
<link rel="alternate stylesheet" media="screen,projection" type="text/css" href="css/1col.css" title="1col" />
<!--[if lte IE 6]><link rel="stylesheet" media="screen,projection" type="text/css" href="css/main-ie6.css" /><![endif]-->
<link rel="stylesheet" media="screen,projection" type="text/css" href="css/style.css" />
<link rel="stylesheet" media="screen,projection" type="text/css" href="css/mystyle.css" />
<script type="text/javascript" src="js/jquery.js"></script>
<script type="text/javascript" src="js/switcher.js"></script>
<script type="text/javascript" src="js/toggle.js"></script>
<script type="text/javascript" src="js/ui.core.js"></script>
<script type="text/javascript" src="js/ui.tabs.js"></script>
<script type="text/javascript">
	$(document).ready(function(){
		$(".tabs > ul").tabs();
	});
	</script>
</head>
<body>
<div id="main">
  <!-- Tray -->
  <div id="tray" class="box">
    <p class="f-left box">
      <!-- Switcher -->
      <span class="f-left" id="switcher"> <a href="javascript:void(0);" rel="1col" class="styleswitch ico-col1" title="Display one column"><img src="design/switcher-1col.gif" alt="1 Column" /></a> <a href="javascript:void(0)" rel="2col" class="styleswitch ico-col2" title="Display two columns"><img src="design/switcher-2col.gif" alt="" /></a> </span><strong>OptiMAS: a decision support tool to conduct marker-assisted recurrent selection programmes</strong> </p>
      <p class="f-right">Last release: <strong>OptiMAS v1.4 (Jan. 24th, 2013)</strong></p>
  </div>
  <!--  /tray -->
  <hr class="noscreen" />
  <!-- Menu -->
  <div id="menu" class="box">
    <ul class="box f-right">
      <li><a href="https://www.integratedbreeding.net/"><span><strong>Visit IBP Site &raquo;</strong></span></a></li>
    </ul>
    <ul class="box">
      <!-- Active -->
      <li><a href="index.html"><span>Home</span></a></li>
      <li><a href="download.html"><span>Download</span></a></li>
      <li><a href="contact.html"><span>Contact</span></a></li>
    </ul>
  </div>
  <!-- /header -->
  <hr class="noscreen" />
  <!-- Columns -->
  <div id="cols" class="box">
    <!-- Aside (Left Column) -->
    <div id="aside" class="box">
      <div class="padding box">
        <!-- Logo (Max. width = 200px) -->
        <p id="logo"><img src="tmp/logo_optimas.png" alt="" /></p>
      </div>
      <!-- /padding -->
      <ul class="box">
		<li class="submenu-active"><a>OptiMAS</a></li>
        <li><a href="index.html">Homepage</a></li>
        <li><a href="download.html">Download</a></li>
		<li><a href="screenshots.html">Screenshots</a></li>
		<li><a href="citation.html">Papers &amp; Presentations</a></li>
		<li><a href="contact.html">Contact &amp; Support</a></li>
		<li><a href="news.html">What's New</a></li>
		<li class="submenu-active"><a>Online Help</a></li>
		<li><a href="install.html"><strong>Installation procedure</strong></a>
	  		<ul>
            	<li><a href="install_cmd.html">OptiMAS in command line</a></li>
            	<li><a href="install_gui.html">OptiMAS GUI</a></li>
	  		</ul>
		</li>
        <li><a href="data_preparation.html"><strong>Data preparation</strong></a>
	  		<ul>
            	<li><a href="genotype.html">Genotypes/Pedigree file (.dat)</a></li>
				<li><a href="map.html">Genetic Map (.map)</a></li>
	  		</ul>
		</li>
		<li><a href="optimas_cmd.html"><strong>OptiMAS in command line</strong></a>
	  		<ul>
            	<li><a href="run_cmd.html">Running OptiMAS in cmd line</a></li>
				<li><a href="res_output.html">Results &amp; Output files</a></li>
	  		</ul>
		</li>
		<li><a><strong>OptiMAS GUI</strong></a>
	  		<ul>
            	<li><a href="run_gui.html">Running OptiMAS GUI</a></li>
				<li><a href="step1.html">Step1 - Prediction</a></li>
				<li><a href="step2.html">Step2 - Selection</a></li>
				<li><a href="step3.html">Step3 - Intermating</a></li>
				<li><a href="save_reload.html">Saving/Reloading</a></li>
				<li><a href="tools.html">Tools</a></li>
	  		</ul>
		</li>
	  </ul>
	<p class="centrer"><img src="tmp/valid-xhtml10.png" alt="Valid XHTML 1.0 Strict" height="31" width="88" /><img style="border:0;width:88px;height:31px" src="tmp/vcss-blue.gif" alt="CSS Valide !" /></p>
	<p class="centrer"><a href="http://www.inra.fr"><img src="tmp/logo_inra.png" alt="logo_inra" /></a><a href="https://www.integratedbreeding.net/"><img src="tmp/logo_ibp.png" alt="logo_ibp" /></a></p>
	<p class="centrer"><a href="http://www.generationcp.org/"><img src="tmp/logo_gcp.png" alt="logo_gcp" /></a><a href="http://moulon.inra.fr"><img src="tmp/logo_moulon.png" alt="logo_moulon" /></a></p>
    </div>
    <!-- /aside -->
    <hr class="noscreen" />
    <!-- Content (Right Column) -->
    <div id="content" class="box">
      <h1>Data preparation - Genotypes/Pedigree file (.dat) </h1>
	  <p>The genotypes/pedigree file should contain individuals, pedigree information and genotypic data. It also requires a header line specifying the information in each column. The file organization will be exemplified below for the following MAS pedigree.</p>
	  <p><img src="tmp/fig2.png" alt="Figure 2: example of a multiparental design"><br/><b><i>Figure 2: example of a multiparental design</i></b></p>
	  <p>The default structure for the input file (corresponding to Fig. 2 above) can be represented as follows:</p>
	  <!-- Table (TABLE) -->
      <table>
        <tr><th>Id</th><th><b class="red">P1</b></th><th><b class="red">P2</b></th><th><b class="red">Step</b></th><th><b class="red">Cycle</b></th><th><b class="red">Group</b></th><th>Mk1</th><th>Mk2</th><th>Mk3</th><th>...</th><td></td></tr>
	    <tr><td>IL1</td><td>a</td><td>a</td><td>IL</td><td>IL</td><td>-</td><td>T</td><td>C</td><td>A</td><td>...</td><td><b>[1]</b></td></tr>
		<tr class="bg"><td>IL2</td><td>b</td><td>b</td><td>IL</td><td>IL</td><td>-</td><td>C</td><td>C</td><td>G</td><td>...</td></tr>
		<tr><td>IL3</td><td>c</td><td>c</td><td>IL</td><td>IL</td><td>-</td><td>T</td><td>C</td><td>A</td><td>...</td></tr>
		<tr class="bg"><td>IL4</td><td>d</td><td>d</td><td>IL</td><td>IL</td><td>-</td><td>T</td><td>C</td><td>G</td><td>...</td></tr>
		<tr><td>IL1_IL2_F1</td><td>IL1</td><td>IL2</td><td>CR</td><td>F1</td><td></td><td></td><td></td><td></td><td>...</td><td><b>[2]</b></td></tr>
		<tr class="bg"><td>IL3_IL4_F1</td><td>IL3</td><td>IL4</td><td>CR</td><td>F1</td><td></td><td></td><td></td><td></td><td>...</td></tr>
		<tr><td>IL1_IL2_F2</td><td>IL1_IL2_F1</td><td>IL1_IL2_F1</td><td>S1</td><td>F2</td><td>-</td><td>T</td><td>C</td><td>A</td><td>...</td><td><b>[3]</b></td></tr>
		<tr class="bg"><td>IL3_IL4_F2</td><td>IL3_IL4_F1</td><td>IL3_IL4_F1</td><td>S1</td><td>F2</td><td>-</td><td>T</td><td>C</td><td>G/A</td><td>...</td></tr>
		<tr><td>Ind1</td><td>IL1_IL2_F2</td><td>IL3_IL4_F2</td><td>CR</td><td>C1</td><td>Ear1</td><td>T</td><td>-</td><td>G/A</td><td>...</td><td><b>[4]</b></td></tr>
		<tr class="bg"><td>Ind2</td><td>IL1_IL2_F2</td><td>IL3_IL4_F2</td><td>CR</td><td>C1</td><td>Ear2</td><td></td><td></td><td></td><td>...</td><td><b>[5]</b></td></tr>
      </table>
	  <p><b>The genotype/pedigree file must be in plain-text, tab-delimited format (no space between fields). The header should not be changed even if the 2 optional columns (Cycle/Group) are left blank (or "-").</b></p>
	  <p>After the header line, each line contains the name (Id) of the current individual followed by its parents (P1/P2), the pedigree relationship linking the two generations (Step), assignation to cycle and group (if relevant) and the genotyping data (Mk1-Mkn). Note that this format is very close to the input format of the <a href="http://bioinf.scri.ac.uk/flapjack/index.shtml">Flapjack</a> software except that five additional columns (in red) must be added.</p>
	  <p><u>Columns of this file correspond to</u>:</p>
	  <p><b>Id:</b> corresponds to the name of each individual coded as a character string without any blanks and special characters. It must be unique.</p>
	  <p><b>Parent 1, Parent 2:</b> correspond to the name(s) of the parent(s) of the individual (must exist as individuals above in the file except for founder parents). Individuals must be ranked according to generations (from oldest to most recent).<br/>
			Founder parents of the programme ([1] in table) are assumed to be homozygous lines with no residual heterozygosity. Their pedigree is assumed to be unknown. Parent1, parent2 columns indicate in this case the allele that will be transmitted through generations. This allele is identified as a character strain up to 6 characters. OptiMAS can also handle selection schemes starting from heterozygous individuals (e.g. fruit tree breeding). In this case two virtual inbred lines must be defined for each heterozygous founder (described then as a Cross, see below).</p>
	  <p><b>Step</b>: corresponds to the pedigree relationship between the individual and its parent(s):<br/>
	  <b class="marge">CR (Cross):</b> indicates that the individual results from a cross between its two parents.<br/>
	  <b class="marge">Sn:</b> indicates that the individual results from n (integer between 1 and 20) generations of selfing of its parent (in this case the two parents Id must be identical).<br/>
	  <b class="marge">RIL:</b> Recombinant Inbred Lines (assumes that the individual results from an infinite number of selfing generations from an initial F1 hybrid).<br/>
	  <b class="marge">DH:</b> Double Haploids (assumes that the individual results from haplo-diploïdisation from an initial F1 hybrid).<br/>
	  <b class="marge">IL:</b> indicates founder inbred status (see above).</p>
	  <p><b>Cycle:</b> optional information regarding the generation in the programme (e.g. first cycle, second cycle, F2, F4, etc).</p>
	  <p><b>Group:</b> optional information regarding another classification criterion (e.g. subprogrammes, families, etc).</p>
	  <p><b>Mk1-Mkn:</b> genotyping results. The software can deal with SNPs, microsatellites and any bi/multi allelic marker genotyping technique with either dominant or codominant scoring.<br/> Note that:<br/>
		<span class="marge">The markers present in the genotype/pedigree file must be ordered and match those in the map file (same name and number of markers).<br/></span>
		<span class="marge">Homozygous genotypes for an allele (e.g. A) can be scored either as A or A/A.<br/></span>
		<span class="marge">Heterozygotes are expected to be separated by a "/" (e.g. A/G). Heterozygous genotypes are assumed to be unphased (i.e. A/B equivalent to B/A).<br/></span>
		<span class="marge">Missing data at marker loci are allowed and must be entered as "-" (or can be left blank).<br/></span>
		<span class="marge">For dominant markers, assuming A dominant vs. a recessive, genotypes presenting allele A must be coded A/-.<br/></span>
		<span class="marge">Parental inbred lines should not contain missing data.<br/></span></p>
	  <p><u>Description of the genotypes/pedigree file example, in relationship with the multiparental MARS schema (see table above):</u><br/>
			<b>[1]</b> Parental line where a is the name given to its alleles (same name for all loci).<br/>
			<b>[2]</b> Example of a hybrid F1 obtained by intercrossing two parental lines. In this case, the genotype is inferred from parents and does not have to be declared.<br/>
			<b>[3]</b> Plant issued from a selfing process. The name of the two parents is the same in this case. S1 means that this individual was obtained after one generation of selfing. The number of generation(s) can vary.<br/>
			<b>[4]</b> Plant issued from the cross between two individuals (in this case, two F2 coming from different parental lines) already declared and genotyped. In this case, the four parental alleles may have been transmitted to this individual.<br/> 
			<b>[5]</b> Plant, without genotyping data information, issued from the cross between two F2 individuals. All possible genotypes will be considered to evaluate the genetic value of this plant.</p>
	  <p>Given the <b>possibility to include non genotyped individuals</b> (e.g. Ind2), this makes it possible to analyze most common MAS schemes and mating designs. So, if several (non genotyped) steps were required to obtain a specific individual, we must generate virtual individuals in these intermediate steps.</p>
	  <p>&nbsp;</p>
    </div>
    <!-- /content -->
  </div>
  <!-- /cols -->
  <hr class="noscreen" />
  <!-- Footer -->
  <div id="footer" class="box">
    <p class="f-left">&copy; 2009 <u>OptiMAS</u>, All Rights Reserved &reg;</p>
	<!-- Do you want remove this backlinks? Look at www.nuviotemplates.com/payment/ -->            
    <p class="f-right"><a href="http://www.adminzio.com/">Adminzio</a> is Free <a href="http://www.adminzio.com/">Admin template</a> provided by <a href="http://www.nuviotemplates.com/">Nuvio Templates</a> available for download at <a href="http://www.nuviotemplates.com/admin-templates/">Admin Templates</a></p>
    <!-- Do you want remove this backlinks? Look at www.nuviotemplates.com/payment/ -->
  </div>
  <!-- /footer -->
</div>
<!-- /main -->
</body>
</html>
